### PR TITLE
Adding initial JMX support to Grid Hub

### PR DESCRIPTION
**Pull request by dimacus**
This is a starting point to the integration of JMX into the grid.

Current PR keeps track of the Registry class for new sessions and the RemoteProxy class for new proxy registration, as seen bellow:
java monitoring management console-3

If this is a desired direction, we can add TestSlot tracking etc, and even give ability to temporary offline a node via console or to force clear a TestSlot that is stuck in an un flush-able state.

Directions to test:

```
java -Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.port=9010 -Dcom.sun.management.jmxremote.local.only=false -Dcom.sun.management.jmxremote.authenticate=false -Dcom.sun.management.jmxremote.ssl=false -Djava.rmi.server.hostname=localhost -jar selenium-standalone.jar -role hub
```

(Add a new node in normal fashion)

Next run this command to find the PID of the hub server started

```
jps
```

Once the pid is located, run jconsole command to connect to pid:

```
jconsole $PID
```

(ignore any insecure connection errors you might see)

---
This change is [Reviewable](https://reviewable.io/reviews/seleniumhq/selenium/316)

### CODE REVIEW

**Comment by andreastt** [comment]
I don't know what JMX is, maybe @barancev can take a look?

File: java/server/.classpath [hunk 1]
```diff
@@ -23,5 +23,6 @@
 	<classpathentry kind="lib" path="/third-party/java/netty/netty-3.5.7.Final.jar"/>
 	<classpathentry kind="lib" path="/third-party/java/guava-libraries/guava-18.0.jar" sourcepath="/third-party/java/guava-libraries/guava-18.0-sources.jar"/>
 	<classpathentry kind="lib" path="/third-party/java/mockito/mockito-core-1.9.5.jar"/>
+	<classpathentry kind="lib" path="/third-party/java/jmxutils/jmxutils-1.18.jar" sourcepath="/third-party/java/jmxutils/jmxutils-1.18-sources.jar"/>
```
Comment (shs96c) [inline java/server/.classpath]: Interesting that this works at all. The third party directory is "third_party" 

/me shrugs

File: java/server/src/org/openqa/grid/internal/Registry.java [hunk 2]
```diff
@@ -19,6 +19,8 @@
 
 import com.google.common.base.Predicate;
 
+import com.sun.org.glassfish.gmbal.ManagedObject;
```
Comment (shs96c) [inline java/server/src/org/openqa/grid/internal/Registry.java]: JMX offers normal APIs for this. I don't think you want the glassfish one.

File: java/server/src/org/openqa/grid/internal/Registry.java [hunk 3]
```diff
@@ -176,6 +194,12 @@ public void removeIfPresent(RemoteProxy proxy) {
       log.warning(String.format(
           "Proxy '%s' was previously registered.  Cleaning up any stale test sessions.", proxy));
 
+      // TODO refactor all checks like this and registration
+      if(System.getProperty("com.sun.management.jmxremote") != null) {
```
Comment (shs96c) [inline java/server/src/org/openqa/grid/internal/Registry.java]: Second time you've used this check. Extract into a boolean method: `isJmxEnabled` or something similar

File: java/server/src/org/openqa/grid/internal/TestSession.java [hunk 4]
```diff
@@ -21,6 +21,8 @@
 import com.google.common.io.ByteStreams;
 import com.google.common.net.MediaType;
 
+import com.sun.org.glassfish.gmbal.ManagedObject;
```
Comment (shs96c) [inline java/server/src/org/openqa/grid/internal/TestSession.java]: Again, you don't want to depend on glassfish.

File: java/server/src/org/openqa/grid/internal/TestSlot.java [hunk 5]
```diff
@@ -18,12 +18,19 @@
 
 import com.google.common.base.Throwables;
 
+import com.sun.org.glassfish.gmbal.ManagedObject;
```
Comment (shs96c) [inline java/server/src/org/openqa/grid/internal/TestSlot.java]: And again

File: java/server/src/org/openqa/grid/web/Hub.java [hunk 6]
```diff
@@ -19,6 +19,8 @@
 
 import com.google.common.collect.Maps;
 
+import com.sun.org.glassfish.gmbal.ManagedObject;
```
Comment (shs96c) [inline java/server/src/org/openqa/grid/web/Hub.java]: And again. The reason it's bad is that if someone uses a JDK not produced by Oracle they won't have this class.

Comment (gblikegroupon) [inline java/server/src/org/openqa/grid/web/Hub.java]: These comments were just pointed out to me. This is a pretty easy fix. I'll make the appropriate updates to the other imports as well.

File: java/server/src/org/openqa/grid/internal/Registry.java (hunk 3 above)
Comment (gblikegroupon) [inline java/server/src/org/openqa/grid/internal/Registry.java]: Yeah. I'll make that todo a todonow.

**Comment by andreastt** [comment]
@gblikegroupon, do you have time to look at the fixups needed to land this?

**Comment by gblikegroupon** [comment]
@andreastt I had a few high priority items come down the pipe at the same time. I'm working on this right now.

**Comment by gblikegroupon** [comment]
I have made some appropriate changes. Since this pull request is handled through @dimacus's  fork, I need to get hold of him before I can share them. Or if you just want to take a peek at them, https://github.com/gblikegroupon/selenium/tree/jmx.

**Comment by mar1ged** [comment]
Just out of curiousity: don't you think that a more generic approach would be a good alternative ? JMX often is there for monitoring purposes. In order to work around the problems created by JMX (security etc) often we use Jolokia to expose the information.
And in the end (at least in the company I work at) we only use all of this to create metrics which we will pull into monitoring systems like Prometheus.

Would you consider creating a Prometheus integration a good alternative ?

**Comment by barancev** [comment]
This PR was not completed. Also JMX support was added to the Grid independently on this PR, so it is not actual anymore. Anyway, thank you for the effort and inspiration!

**Pull request closed without merging**